stackslib/src/net/download/nakamoto.rs

```diff
@@ -0,0 +1,4283 @@
+        // tenure ID of the ongoing tenure.  It corresponds to the highest sortition for which
+        // there exists a tenure.
+        //
+        // There are three possibilities for obtaining this, based on what we know about tenures
+        // from the sortition DB and the peers' inventories:
+        //
+        // Case 1: There are no sortitions yet in the current reward cycle, so this is the
+        // second-to-last WantedTenure in the last reward cycle's WantedTenure list.
+        //
+        // Case 2: There is one sortition in the current reward cycle, so this is the last
+        // WantedTenure in the last reward cycle's WantedTenure list
+        //
+        // Case 3: There are two or more sortitions in the current reward cycle, so this is the
+        // second-to-last WantedTenure in the current reward cycle's WantedTenure list.
+        let highest_wanted_tenure = if self.wanted_tenures.len() == 0 {
+            // highest complete wanted tenure is the second-to-last tenure in prev_wanted_tenures
+            let Some(prev_wanted_tenures) = self.prev_wanted_tenures.as_ref() else {
+                // not initialized yet (technically unrachable)
+                return HashMap::new();
+            };
+            if prev_wanted_tenures.len() < 2 {
+                return HashMap::new();
+            };
+            let Some(wt) = prev_wanted_tenures.get(prev_wanted_tenures.len().saturating_sub(2))
+            else {
+                return HashMap::new();
+            };
+            wt.clone()
+        } else if self.wanted_tenures.len() == 1 {
+            // highest complete tenure is the last tenure in prev_wanted_tenures
+            let Some(prev_wanted_tenures) = self.prev_wanted_tenures.as_ref() else {
+                return HashMap::new();
+            };
+            let Some(wt) = prev_wanted_tenures.last() else {
+                return HashMap::new();
+            };
+            wt.clone()
+        } else {
+            // highest complete tenure is the second-to-last tenure in wanted_tenures
+            let Some(wt) = self
+                .wanted_tenures
+                .get(self.wanted_tenures.len().saturating_sub(2))
+            else {
+                return HashMap::new();
+            };
+            wt.clone()
+        };
+
+        // Run the confirmed downloader state machine set, since we could already be processing the
+        // highest complete tenure download.  NOTE: due to the way that we call this method, we're
+        // guaranteed that if the `tenure_downloads` downloader set has any downloads at all, they
+        // will only be for the highest complete tenure (i.e. we only call this method if we've
+        // already downloaded all confirmed tenures), so there's no risk of clobberring any other
+        // in-flight requests.
+        let new_confirmed_blocks = if self.tenure_downloads.inflight() > 0 {
```

Comment:
Prob not worth it, but we have a number of inflight() > 0 checks. Might be worth having a helper function called "has_inflight" which wraps this to return the bool.